### PR TITLE
Exclude pkg main files from `no-unused-modules`

### PR DESCRIPTION
**Pull request by rfermann**
Implements a solution for #1327

### CODE REVIEW

File: docs/rules/no-unused-modules.md [hunk 1]
```diff
@@ -105,7 +105,8 @@ export function doAnything() {
 export default 5 // will not be reported
 ```
 
-
+#### Important Note
+Exports from files listed as a main file (`main`, `browser` or `bin` fields in package.json) will be ignored by default. This does only apply, if the package.json is not set to `private: true`
```
Comment (ljharb) [inline docs/rules/no-unused-modules.md]:
```suggestion
Exports from files listed as a main file (`main`, `browser`, or `bin` fields in package.json) will be ignored by default. This does only apply, if the package.json is not set to `private: true`
```

File: src/rules/no-unused-modules.js [hunk 2]
```diff
@@ -201,6 +203,42 @@ const newNamespaceImportExists = specifiers =>
...
+  }
+
+  if (pkg.bin && typeof pkg.bin === 'object') {
+    const binFiles = Object.values(pkg.bin).map(value => join(basePath, value))
+    if (binFiles.includes(file)) {
+      return true
+    }
+  }
+
+  if (pkg.browser) {
```
Comment (ljharb) [inline src/rules/no-unused-modules.js]: The browser field can also be an object.

Comment (rfermann) [reply src/rules/no-unused-modules.js]: Thanks for this info. This wasn't mentioned in the [npm documentation](https://docs.npmjs.com/files/package.json). I will quickly add this part.

Comment (ljharb) [reply src/rules/no-unused-modules.js]: https://github.com/defunctzombie/package-browser-field-spec

File: src/rules/no-unused-modules.js [hunk 3]
```diff
@@ -201,6 +203,49 @@ const newNamespaceImportExists = specifiers =>
...
 
+const fileIsInPkg = file => {
+  const { path, pkg } = readPkgUp.sync({cwd: file, normalize: false})
+  const basePath = dirname(path)
+
+  if (pkg.private === true) {
+    return false
+  }
+
+  if (pkg.bin && typeof pkg.bin === 'string') {
```
Comment (ljharb) [inline src/rules/no-unused-modules.js]: seems like we can consolidate the logic for “check string field X” and “check object field Y”, so t doesn’t have to be repeated for main/bin/browser and bin/browser?

Comment (rfermann) [reply src/rules/no-unused-modules.js]: Good hint. I mad the logic a bit more generic with 3ed222d.

File: src/rules/no-unused-modules.js [hunk 4]
```diff
@@ -207,38 +207,39 @@ const fileIsInPkg = file => {
...
-    if (browserFiles.includes(file)) {
+  if (pkg.browser) {
+    if (checkPkgField(pkg.browser)) {
       return true
     }
   }
 
   if (pkg.main) {
-    if (join(basePath, pkg.main) === file) {
+    if (checkPkgField(pkg.main)) {
```
Comment (ljharb) [inline src/rules/no-unused-modules.js]: the only remaining problem here is that main can’t be an object :-/

Comment (rfermann) [reply src/rules/no-unused-modules.js]: Maybe I don't see the issue here. `main` will either return `true` within the first `if` branch or just return at the end of the function.
Would it be better to return `false` in the first `if` branch to make `main` not even pass the `if` branch for object handling?

Comment (ljharb) [reply src/rules/no-unused-modules.js]: Yes, that's what I'm suggesting. an object "main" field is just invalid and shouldn't be considered.

File: docs/rules/no-unused-modules.md [hunk 5]
```diff
@@ -105,7 +105,8 @@ export function doAnything() {
 export default 5 // will not be reported
 ```
 
-
+#### Important Note
+Exports from files listed as a main file (`main`, `browser`, or `bin` fields in package.json) will be ignored by default. This does only apply, if the package.json is not set to `private: true`
```
Comment (ljharb) [inline docs/rules/no-unused-modules.md]:
```suggestion
Exports from files listed as a main file (`main`, `browser`, or `bin` fields in `package.json`) will be ignored by default. This only applies if the `package.json` is not set to `private: true`
```

File: tests/files/no-unused-modules/main/index.js [hunk 6]
```diff
@@ -0,0 +1 @@
+export const main = 'main'
```
Comment (ljharb) [inline tests/files/no-unused-modules/main/index.js]: please ensure every file has a trailing newline

Comment (rfermann) [reply tests/files/no-unused-modules/main/index.js]: I always miss that. In my other projects it gets added automatically by ESLint/Prettier...
Can we probably add `eol-last` to the eslint-config of this project?

Comment (ljharb) [reply tests/files/no-unused-modules/main/index.js]: Sure, sounds great.

Comment (ljharb) [reply tests/files/no-unused-modules/main/index.js]: (but, your editor should add a trailing newline by default in every file, entirely unrelated to eslint)

File: src/rules/no-unused-modules.js [hunk 7]
```diff
@@ -201,6 +203,58 @@ const newNamespaceImportExists = specifiers =>
...
+  const basePath = dirname(path)
+
+  const checkPkgFieldString = pkgField => {      
+    if (join(basePath, pkgField) === file) {
+        return true
+      }
+  }
+
+  const checkPkgFieldObject = pkgField => {    
+      const pkgFieldFiles = Object.values(pkgField).map(value => join(basePath, value))
```
Comment (ljharb) [inline src/rules/no-unused-modules.js]: `Object.values` is not available on all supported nodes; please use https://npmjs.com/object.values instead

Comment (rfermann) [reply src/rules/no-unused-modules.js]: I would rather use `_.values(object)`, if it is okay for you. Lodash already is a project dependency, with object.values we would add a new dependency.

Comment (ljharb) [reply src/rules/no-unused-modules.js]: I'd prefer to use the standard method; adding a dependency isn't a problem.

File: src/rules/no-unused-modules.js [hunk 8]
```diff
@@ -201,6 +204,58 @@ const newNamespaceImportExists = specifiers =>
...
+
+  const checkPkgFieldString = pkgField => {
+    if (join(basePath, pkgField) === file) {
+        return true
+      }
+  }
+
+  const checkPkgFieldObject = pkgField => {
+      const pkgFieldFiles = values(pkgField).map(value => join(basePath, value))
+      if (pkgFieldFiles.includes(file)) {
```
Comment (ljharb) [inline src/rules/no-unused-modules.js]: includes also isn't in node 4; i'll update the PR with a dep for that too.

**Pull request merged**